agent/functions/put_file_test.go

```diff
@@ -0,0 +1,193 @@
+	}{
+		{
+			name: "正常系：新規ファイル作成",
+			input: PutFileInput{
+				OutputPath:  "test.txt",
+				ContentText: "test content",
+			},
+			checkFile: true,
+			validateFunc: func(t *testing.T, got store.File, err error) {
+				if err != nil {
+					t.Errorf("unexpected error: %v", err)
+				}
+				if got.Path != "test.txt" || got.Content != "test content\n" {
+					t.Errorf("unexpected result: got %+v", got)
+				}
+			},
+		},
+		{
+			name: "正常系：ディレクトリ作成を伴うファイル作成",
+			input: PutFileInput{
+				OutputPath:  "new/dir/test.txt",
+				ContentText: "nested content",
+			},
+			checkFile: true,
+			validateFunc: func(t *testing.T, got store.File, err error) {
+				if err != nil {
+					t.Errorf("unexpected error: %v", err)
+				}
+				if got.Path != "new/dir/test.txt" || got.Content != "nested content\n" {
+					t.Errorf("unexpected result: got %+v", got)
+				}
+			},
+		},
+		{
+			name: "正常系：既存ファイルの上書き",
+			input: PutFileInput{
+				OutputPath:  "existing.txt",
+				ContentText: "new content",
+			},
+			setupFunc: func(t *testing.T, dir string) {
+				path := filepath.Join(dir, "existing.txt")
+				if err := os.WriteFile(path, []byte("old content"), 0644); err != nil {
+					t.Fatal(err)
+				}
+			},
+			checkFile: true,
+			validateFunc: func(t *testing.T, got store.File, err error) {
+				if err != nil {
+					t.Errorf("unexpected error: %v", err)
+				}
+				if got.Path != "existing.txt" || got.Content != "new content\n" {
+					t.Errorf("unexpected result: got %+v", got)
+				}
+			},
+		},
+		{
+			name: "異常系：無効なパス（ファイル名なし）",
+			input: PutFileInput{
+				OutputPath:  "/invalid/path/",
+				ContentText: "test content",
+			},
+			wantErr: true,
+			validateFunc: func(t *testing.T, got store.File, err error) {
+				if err == nil {
+					t.Error("expected error but got nil")
+				}
+			},
+		},
+		{
+			name: "異常系：書き込み権限なしディレクトリ",
+			input: PutFileInput{
+				OutputPath:  "noperm/test.txt",
+				ContentText: "test content",
+			},
+			setupFunc: func(t *testing.T, dir string) {
+				noPermDir := filepath.Join(dir, "noperm")
+				if err := os.Mkdir(noPermDir, 0000); err != nil {
+					t.Fatal(err)
+				}
+			},
+			wantErr: true,
+			validateFunc: func(t *testing.T, got store.File, err error) {
+				if err == nil {
+					t.Error("expected error but got nil")
+				}
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// テスト用の一時ディレクトリを作成
+			tmpDir := t.TempDir()
+			
+			// 現在のディレクトリを保存
+			currentDir, err := os.Getwd()
+			if err != nil {
+				t.Fatal(err)
+			}
+			
+			// テストディレクトリに移動
+			if err := os.Chdir(tmpDir); err != nil {
+				t.Fatal(err)
+			}
+			
+			// テスト終了時に元のディレクトリに戻る
+			t.Cleanup(func() {
+				os.Chdir(currentDir)
+			})
+
+			// セットアップ関数の実行
+			if tt.setupFunc != nil {
+				tt.setupFunc(t, tmpDir)
+			}
+
+			// テスト実行
+			got, err := PutFile(tt.input)
+
+			// バリデーション関数の実行
+			if tt.validateFunc != nil {
+				tt.validateFunc(t, got, err)
+			}
```

Comment:
from security_reviewer
競合状態（race condition）に対するテストが不足しています。同じファイルに対する同時アクセスや、TOCTOUの脆弱性に対するテストを追加すべきです。

```suggestion
以下のようなテストケースを追加することを推奨します：
```go
{
    name: "異常系：競合状態の検出",
    input: PutFileInput{
        OutputPath: "race.txt",
        ContentText: "test content",
    },
    setupFunc: func(t *testing.T, dir string) {
        // 同時アクセスをシミュレート
        go func() {
            os.WriteFile("race.txt", []byte("concurrent write"), 0644)
        }()
    },
    wantErr: true,
},
```
```